### ISSUE TITLE
Presence of "*.partial" files in the "pkgs" directory cause md5 mismatch errors

### ISSUE THREAD

**Issue by nicoddemus** [conda/conda#13488]
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

We recently started using `conda-lock` and started to eventually get errors like this on CI:

```
09:11:22 ERROR:root:ChecksumMismatchError: Conda detected a mismatch between the expected content and downloaded content
09:11:22 ERROR:root:for url 'https://eden.esss.co/conda-channel/mirror-conda-forge/win-64/libogg-1.3.4-h8ffe710_1.tar.bz2'.
09:11:22 ERROR:root:  download saved to: W:\Miniconda\pkgs\libogg-1.3.4-h8ffe710_1.tar.bz2
09:11:22 ERROR:root:  expected md5: 04286d905a0dcb7f7d4a12bdfe02516d
09:11:22 ERROR:root:  actual md5: 721ab132d4b3cb8f3f17e040557bd22e
09:11:22 ERROR:root:
09:11:22 ERROR:root:ChecksumMismatchError: Conda detected a mismatch between the expected content and downloaded content
09:11:22 ERROR:root:for url 'https://eden.esss.co/conda-channel/mirror-conda-forge/win-64/libogg-1.3.4-h8ffe710_1.tar.bz2'.
09:11:22 ERROR:root:  download saved to: W:\Miniconda\pkgs\libogg-1.3.4-h8ffe710_1.tar.bz2
09:11:22 ERROR:root:  expected md5: 04286d905a0dcb7f7d4a12bdfe02516d
09:11:22 ERROR:root:  actual md5: 721ab132d4b3cb8f3f17e040557bd22e
```

It was reproducible on the same machine, but the packages would differ every time.

We did some digging, but to summarize:

* We made a backup of `W:\Miniconda\pkgs` and then removed it, the problem went away.
* We restored the backup, problem started to occur again.
* Noticed that there were a lot of `*.partial` files in `W:\Miniconda\pkgs`, so we removed them -- problem went away.

We will implement a workaround for now to remove `*.partial` files from `W:\Miniconda\pkgs` at the start of the CI jobs, should fix that problem (will come back in case it does not for some other reason).

We only seen this problem on Windows.

While the problem occurs using `conda-lock`, we figure it uses the `conda` infrastructure to download the files, so thought it was appropriate to report this here -- let me know if that's not the case and I will report this in `conda-lock`.

Thanks!

### Conda Info

```shell
active environment : eden-py310
    active env location : W:\Souring\envs\eden-py310
            shell level : 1
       user config file : C:\Users\bruno\.condarc
 populated config files : W:\Miniconda\.condarc
                          C:\Users\bruno\.condarc
                          W:\Souring\.condarc
          conda version : 23.11.0
    conda-build version : not installed
         python version : 3.10.2.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=x86_64
                          __conda=23.11.0=0
                          __cuda=11.7=0
                          __win=0=0
       base environment : W:\Souring\envs\_eden-conda-base.2  (writable)
      conda av data dir : W:\Souring\envs\_eden-conda-base.2\etc\conda
  conda av metadata url : None
           channel URLs : https://@eden.esss.co/conda-channel/esss/win-64
                          https://@eden.esss.co/conda-channel/esss/noarch
                          https://@eden.esss.co/conda-channel/mirror-main/win-64
                          https://@eden.esss.co/conda-channel/mirror-main/noarch
                          https://@eden.esss.co/conda-channel/mirror-conda-forge/win-64
                          https://@eden.esss.co/conda-channel/mirror-conda-forge/noarch
          package cache : W:\Miniconda\pkgs
       envs directories : W:\Souring\envs
                          w:\Souring\envs
                          W:\Souring\envs\_eden-conda-base.2\envs
                          C:\Users\bruno\.conda\envs
                          C:\Users\bruno\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/23.11.0 requests/2.31.0 CPython/3.10.2 Windows/10 Windows/10.0.22621 solver/libmamba conda-libmamba-solver/23.12.0 libmambapy/1.5.6
          administrator : False
             netrc file : None
           offline mode : False
```
```


### Conda Config

```shell
==> W:\Miniconda\.condarc <==
allow_softlinks: False
create_default_packages:
  - pip
  - python
verify_threads: 4
execute_threads: 4
extra_safety_checks: False
path_conflict: warn
channel_alias: https://@eden.esss.co/conda-channel/
channels:
  - https://@eden.esss.co/conda-channel/esss
  - https://@eden.esss.co/conda-channel/mirror-main
  - https://@eden.esss.co/conda-channel/mirror-conda-forge
show_channel_urls: True
always_yes: True
report_errors: False

==> C:\Users\bruno\.condarc <==
allow_softlinks: False
create_default_packages:
  - pip
  - python
verify_threads: 4
execute_threads: 4
extra_safety_checks: False
path_conflict: warn
channel_alias: https://@eden.esss.co/conda-channel/
channels:
  - https://@eden.esss.co/conda-channel/esss
  - https://@eden.esss.co/conda-channel/mirror-main
  - https://@eden.esss.co/conda-channel/mirror-conda-forge
show_channel_urls: True
always_yes: True
report_errors: False
solver: libmamba

==> W:\Souring\.condarc <==
allow_softlinks: False
create_default_packages:
  - pip
  - python
verify_threads: 4
execute_threads: 4
extra_safety_checks: False
path_conflict: warn
envs_dirs:
  - W:\Souring\envs
  - W:\Souring\\envs
  - w:\Souring\\envs
pkgs_dirs:
  - W:\Miniconda\pkgs
channel_alias: https://@eden.esss.co/conda-channel/
channels:
  - https://@eden.esss.co/conda-channel/esss
  - https://@eden.esss.co/conda-channel/mirror-main
  - https://@eden.esss.co/conda-channel/mirror-conda-forge
show_channel_urls: True
always_yes: True
report_errors: False

==> envvars <==
envs_path:
  - W:\Souring\envs
```
```


### Conda list

```shell
λ conda list --show-channel-urls
# packages in environment at W:\Miniconda:
#
# Name                    Version                   Build  Channel
beautifulsoup4            4.11.1             pyha770c72_0    mirror-conda-forge
brotlipy                  0.7.0           py36h68aa20f_1001    mirror-conda-forge
bzip2                     1.0.6                    vc14_1  [vc14]  mirror-conda-forge
ca-certificates           2023.5.7             h56e8100_0    mirror-conda-forge
certifi                   2021.5.30        py36ha15d459_0    mirror-conda-forge
cffi                      1.14.6           py36he58ceb7_1    mirror-conda-forge
chardet                   4.0.0            py36ha15d459_1    mirror-conda-forge
charset-normalizer        2.1.1              pyhd8ed1ab_0    mirror-conda-forge
conda                     4.11.0+1                 py36_0    esss
conda-build               3.21.9                   py36_1    esss
conda-devenv              1.1.3+2                    py_0    esss
conda-execute             0.9.0                      py_0    mirror-conda-forge
conda-package-handling    1.7.3            py36hb5e345e_0    mirror-conda-forge
console_shortcut          0.1.1                         4    defaults
cryptography              35.0.0           py36hd0de82c_0    mirror-conda-forge
filelock                  3.4.1              pyhd8ed1ab_0    mirror-conda-forge
glob2                     0.7                        py_0    mirror-conda-forge
idna                      2.9                        py_1    defaults
jinja2                    3.0.3              pyhd8ed1ab_0    mirror-conda-forge
libarchive                3.3.3             h0005e80_1004    mirror-conda-forge
libiconv                  1.16                 he774522_0    mirror-conda-forge
liblief                   0.11.5               h0e60522_0    mirror-conda-forge
libxml2                   2.9.12               hf5bbc77_0    mirror-conda-forge
lz4-c                     1.8.3             he025d50_1001    mirror-conda-forge
lzo                       2.10              he774522_1000    mirror-conda-forge
m2-msys2-runtime          2.5.0.17080.65c939c               3    mirror-conda-forge
m2-patch                  2.7.5                         2    mirror-conda-forge
m2w64-gcc-libgfortran     5.3.0                         6    mirror-conda-forge
m2w64-gcc-libs            5.3.0                         7    mirror-conda-forge
m2w64-gcc-libs-core       5.3.0                         7    mirror-conda-forge
m2w64-gmp                 6.1.0                         2    mirror-conda-forge
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    mirror-conda-forge
markupsafe                2.0.1            py36h68aa20f_0    mirror-conda-forge
menuinst                  1.4.18           py36ha15d459_0    mirror-conda-forge
msys2-conda-epoch         20160418                      1    mirror-conda-forge
mu_repo                   1.8.2              pyhd8ed1ab_0    mirror-conda-forge
openssl                   1.1.1l               h8ffe710_0    mirror-conda-forge
pip                       22.0.2             pyhd8ed1ab_0    mirror-conda-forge
pkginfo                   1.8.3              pyhd8ed1ab_0    mirror-conda-forge
powershell_shortcut       0.0.1                         3    defaults
psutil                    5.8.0            py36h68aa20f_1    mirror-conda-forge
py-lief                   0.11.5           py36he2d232f_0    mirror-conda-forge
pycosat                   0.6.3           py36h68aa20f_1006    mirror-conda-forge
pycparser                 2.20                       py_0    defaults
pyopenssl                 22.0.0             pyhd8ed1ab_1    mirror-conda-forge
pysocks                   1.7.1            py36ha15d459_3    mirror-conda-forge
python                    3.6.11          h39d44d4_3_cpython    mirror-conda-forge
python-libarchive-c       3.1              py36ha15d459_0    mirror-conda-forge
python_abi                3.6                     2_cp36m    mirror-conda-forge
pytz                      2022.6             pyhd8ed1ab_0    mirror-conda-forge
pywin32                   301              py36h68aa20f_0    mirror-conda-forge
pyyaml                    5.1.2            py36hfa6e2cd_0    mirror-conda-forge
requests                  2.28.1             pyhd8ed1ab_0    mirror-conda-forge
ripgrep                   13.0.0               h7f3b576_2    mirror-conda-forge
ruamel_yaml               0.15.71         py36hfa6e2cd_1000    mirror-conda-forge
setuptools                58.0.4           py36ha15d459_2    mirror-conda-forge
six                       1.16.0             pyh6c4a22f_0    mirror-conda-forge
soupsieve                 2.3.2.post1        pyhd8ed1ab_0    mirror-conda-forge
sqlite                    3.31.1               h2a8f88b_1    defaults
tqdm                      4.46.0                     py_0    defaults
urllib3                   1.26.11            pyhd8ed1ab_0    mirror-conda-forge
vc                        14.1                 h0510ff6_4    defaults
vs2015_runtime            14.16.27012          hf0eaf9b_1    defaults
wheel                     0.37.1             pyhd8ed1ab_0    mirror-conda-forge
win_inet_pton             1.1.0            py36ha15d459_2    mirror-conda-forge
ww                        1.2.8                         0    esss
xz                        5.2.5                h62dcd97_1    mirror-conda-forge
yaml                      0.1.7                hc54c509_2    defaults
zlib                      1.2.11               h62dcd97_4    defaults
zstd                      1.3.3                    vc14_1  [vc14]  mirror-conda-forge
```

**Comment by nicoddemus** [comment]
Took a glance at the code:

https://github.com/conda/conda/blob/1836e9541b09db65d71cd0257f9ba2b809bfb141/conda/gateways/connection/download.py#L160-L172

I don't have the full picture where this is used, but wouldn't make sense to remove partial file if it does exist before starting the download? If for some reason a `.partial` was left incomplete (for example a CI worker was killed while package download was happening), this would result in a corrupted file, because the file is opened with `a+b`. On the other hand, the docstring does say "Create or open locked partial download file", so seems like opening with `a+b` is intentional?

**Comment by travishathaway** [comment]
Hi @nicoddemus,

Thanks for reporting this. I believe this is the correct place for this issue.

I'm going to tag @dholth here as he was the one primarily responsible for implementing this feature.

**Comment by dholth** [comment]
The `.partial` file exists so that an interrupted conda command can leave it there, and a new conda command can complete the download. If the beginning of the `.partial` file plus the continued download doesn't match the checksum, that's a bug. Is the locking working?

**Comment by nicoddemus** [comment]
> The .partial file exists so that an interrupted conda command can leave it there, and a new conda command can complete the download. 

In which situations a command can be interrupted? 

> Is the locking working?

Which locking you mean?

---

I did take the code from `download.py` for a spin to check if our internal server was responding correctly, and seems like it: it returns `206` for partial content as expected, and I managed to download a file in two "parts", first 100 bytes then the rest, and compare that with a complete download and they match correctly. Did this just to ensure is not some unexpected behavior from our server. I copied the code from `download.py` and adapted it just to hard-code URLs and etc, but did the exact same calls.

**Comment by dholth** [comment]
The main use case is when your slower internet connection is interrupted, conda will time out and exit. Then you should be able to continue without having to re-dowload the beginning of the file.

**Comment by nicoddemus** [comment]
> Conda locks the .partial file during a download so that concurrent conda processes don't write to the same file.

I see, I don't think this is the case here, because our CI workers only run one job at a time, and each time only runs one conda process at a time too.

**Comment by nicoddemus** [comment]
Seems like the underlying code is correct AFAICT.

I see two options:

1. We add this feature behind a feature flag (enabled by default).
2. We close this issue as "won't fix".

I'm OK with 2) given it seems nobody else is hitting this. We can reopen in the future in case more people run into this and/or more information is found.

**Comment by igortg** [comment]
Just to add that I'm also seeing this problem on all Windows nodes of our build farm.
- conda 23.11.0
- conda-lock  2.5.2

**Comment by dholth** [comment]
Could we catch the checksum error and re-download from scratch if there was a mismatch?

**Comment by nicoddemus** [comment]
> Could we catch the checksum error and re-download from scratch if there was a mismatch?

Sounds reasonable... and when that happens, if another mismatch ocurrs, raise an error. 👍

**Comment by ctcjab** [comment]
> 2. We close this issue as "won't fix".
> I'm OK with 2) given it seems nobody else is hitting this. We can reopen in the future in case more people run into this and/or more information is found.

FWIW, a large number of users at my org have been hitting this issue, using the latest version of conda on Linux. It seems to be more likely to occur when using `conda-lock`, which internally calls `conda create` with an `@EXPLICIT` lockfile. And users outside my org are hitting this too, as demonstrated by @romulorosa opening https://github.com/conda/conda-lock/issues/748 last week.

This seems like a case of some faulty optimization logic leading to broken behavior (conda can no longer create environments from lockfiles), and therefore a release should be made with the optimization disabled until it can be reintroduced with more confidence that it won't break anything. Does that make sense?

**Comment by nicoddemus** [comment]
@dholth [suggestion](https://github.com/conda/conda/issues/13488#issuecomment-2296724342) would be a safer alternative too.

**Comment by dholth** [comment]
It's odd that the .tar.bz2 and .conda hashes would get mixed up. We generally treat them as equivalent; I noticed we do include the whole filename in the `.partial` `pytorch-2.2.0-gpu_mps_py310h87e4ab7_100.conda.partial`

**Comment by dholth** [comment]
Is the pkgs directory on a shared or network drive that doesn't support OS locking?

**Comment by nicoddemus** [comment]
In our case no, just a normal directory.

**Comment by ctcjab** [comment]
I have created a repro for this bug here: https://github.com/ctcjab/conda-bug-13488-repro

Curious if anyone who hasn't tried to reproduce this yet is able to now, and hope this helps get to the bottom of the issue.

If you are able to reproduce, would you please consider publishing a hot fix release with the buggy download-resumption optimization disabled, until this bug is fixed and it can be reintroduced?

**Comment by ctcjab** [comment]
Hi all, would anyone be willing to say if you can reproduce this via the repro I provided in my previous message? I think it should take less than 5 minutes, in case anyone can spare the time.

**Comment by dholth** [comment]
I was able to reproduce, it manages to have a `.partial` that's not a prefix of the complete file.
```
z = Path('zipp-3.20.2-pyhd8ed1ab_0.conda').read_bytes()
z2 = Path('zipp-3.20.2-pyhd8ed1ab_0.conda.partial').read_bytes()
z.startswith(z2)
```

**Comment by ctcjab** [comment]
@dholth Thanks so much for confirming you can reproduce! And updating the status here -- great to see it's in progress now and much appreciate your work on it. And finally thank you for the reminder that conda-lock will want to update its vendored version of conda once this is fixed! (/cc @maresb, I'm guessing you're watching this issue already but just in case:)

**Comment by maresb** [comment]
Thanks @ctcjab for staying on top of this. I just took another pass to see if there's anything that I can do here, and I believe the answer is no. Regarding the Conda version that we vendor, we only really use it for Conda's MatchSpec language, and for everything else we invoke a `conda` (or `mamba` or `micromamba`) subprocess.

In particular, the `conda create` command is generated [here](https://github.com/conda/conda-lock/blob/9766789694494c0804460208563e33bf50030934/conda_lock/conda_lock.py#L226-L236). Execution is carried out [here](https://github.com/conda/conda-lock/blob/9766789694494c0804460208563e33bf50030934/conda_lock/invoke_conda.py#L111-L120). If you use the version of `conda-lock` from the `main` branch with `--log-level=DEBUG`, then you'll see what it's doing:

```
DEBUG:conda_lock.invoke_conda:Invoking command: /tmp/conda-bug-13488-repro/.pixi/envs/default/bin/conda create --quiet --file repro.txt --yes --name conda-bug-13488-repro
```

So consequently I think this bug is pretty independent of conda-lock, although if it makes sense I'm open to merging some accommodation as per https://github.com/conda/conda-lock/issues/748#issuecomment-2487872014

> If you use the version of `conda-lock` from the `main` branch with `--log-level=DEBUG`

And ugh, this whole situation is quite painful; sorry for that, I've been really strapped for time lately.

**Comment by dholth** [comment]
I looked into conda-lock's source code, the subprocess calls are pretty easy to follow. I didn't see any kind of parallelization (which could potentially corrupt a `.partial` download in the face of locking bugs).

**Comment by dholth** [comment]
The PR is updated to no longer use 'a' mode, which could cause us to write to the wrong part of the file if the size changes; to retry once only if it is a partial download; and with improved tests.
